File: module-server/src/main/java/de/twomartens/wahlrecht/service/NominationService.java
package de.twomartens.wahlrecht.service;

import de.twomartens.wahlrecht.mapper.v1.NominationMapper;
import de.twomartens.wahlrecht.model.db.Candidate;
import de.twomartens.wahlrecht.model.db.Nomination;
import de.twomartens.wahlrecht.model.internal.NominationId;
import de.twomartens.wahlrecht.repository.NominationRepository;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.mapstruct.factory.Mappers;
import org.springframework.stereotype.Service;

@RequiredArgsConstructor
@Service
public class NominationService {

  private final NominationMapper mapper = Mappers.getMapper(NominationMapper.class);
  private final NominationRepository repository;
  private final CandidateService candidateService;
  private Map<NominationId, Nomination> nominations;

  public Nomination getNomination(NominationId id) {
    if (nominations == null) {
      fetchNominations();
    }
    return nominations.get(id);
  }

  public de.twomartens.wahlrecht.model.internal.Nomination getNominationInternal(NominationId id) {
    return mapper.mapToInternal(getNomination(id));
  }

  public Nomination storeNomination(Nomination nomination) {
    if (nomination == null) {
      return null;
    }

    if (nominations == null) {
      fetchNominations();
    }

    NominationId nominationId = new NominationId(nomination.getElectionName(),
        nomination.getPartyAbbreviation(), nomination.getName());
    Nomination existing = nominations.get(nominationId);
    boolean needsUpdate = !nomination.equals(existing);

    if (!needsUpdate) {
      return existing;
    }

    Collection<Candidate> candidates = new ArrayList<>();
    nomination.getCandidates()
        .forEach(candidate -> candidates.add(candidateService.storeCandidate(candidate)));
    if (existing != null) {
      nomination = existing;
    }
    nomination.setCandidates(candidates);
    nomination = repository.save(nomination);
    nominations.put(nominationId, nomination);

    return nomination;
  }

  private void fetchNominations() {
    nominations = repository.findAll().stream()
        .collect(Collectors.toMap(
            n -> new NominationId(n.getElectionName(), n.getPartyAbbreviation(), n.getName()),
            Function.identity()));
  }
}

File: module-server/src/main/kotlin/de/twomartens/wahlrecht/service/NominationService.kt
package de.twomartens.wahlrecht.service

import de.twomartens.wahlrecht.mapper.v1.NominationMapper
import de.twomartens.wahlrecht.model.db.Nomination
import de.twomartens.wahlrecht.model.internal.NominationId
import de.twomartens.wahlrecht.repository.NominationRepository
import org.mapstruct.factory.Mappers
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Service
class NominationService(
    private val repository: NominationRepository
) {
  private val mapper = Mappers.getMapper(NominationMapper::class.java)

  fun getNominationInternal(id: NominationId): Mono<de.twomartens.wahlrecht.model.internal.Nomination> {
    return getNomination(id).map {
      mapper.mapToInternal(it)
    }
  }

  fun storeNomination(new: Nomination?): Mono<Nomination> {
    if (new == null) return Mono.empty()

    return repository.findAndModify(new)
  }

  fun storeNominations(new: Collection<Nomination>): Flux<Nomination> {
    return Flux.merge(new.map { storeNomination(it) })
  }

  private fun getNomination(id: NominationId): Mono<Nomination> {
    return repository.findByElectionNameAndPartyAbbreviationAndName(
        id.electionName,
        id.partyAbbreviation,
        id.name
    )
  }
}